ffi/src/clipboard/mod.rs

```diff
@@ -0,0 +1,126 @@
+pub mod message;
+
+#[diplomat::bridge]
+pub mod ffi {
+
+    use super::{
+        message::ffi::{ClipboardFormatId, ClipboardFormatIterator, ClipboardMessage, OwndFormatDataResponse},
+        FfiClipbarodMessageProxy,
+    };
+    use crate::error::{ffi::IronRdpError, ValueConsumedError};
+    use ironrdp::cliprdr::Client;
+    use windows::Win32::Foundation::HWND;
+
+    #[diplomat::opaque]
+    pub struct WinCliprdr {
+        pub clipboard: ironrdp_cliprdr_native::WinClipboard,
+        pub receiver: std::sync::mpsc::Receiver<ironrdp::cliprdr::backend::ClipboardMessage>,
+    }
+
+    impl WinCliprdr {
+        pub fn new(hwnd: &Hwnd) -> Result<Box<WinCliprdr>, Box<IronRdpError>> {
+            let (sender, receiver) = std::sync::mpsc::channel();
+
+            let proxy = FfiClipbarodMessageProxy { sender };
+
+            // SAFETY: `hwnd` must be a valid window handle
+            let clipboard = unsafe { ironrdp_cliprdr_native::WinClipboard::new(HWND(hwnd.0), proxy) }?;
+
+            Ok(Box::new(WinCliprdr { clipboard, receiver }))
+        }
+
+        pub fn next_clipboard_message(&self) -> Option<Box<ClipboardMessage>> {
+            self.receiver.try_recv().ok().map(ClipboardMessage).map(Box::new)
+        }
+
+        pub fn backend_facotry(&self) -> Box<CliprdrBackendFactory> {
+            Box::new(CliprdrBackendFactory(self.clipboard.backend_factory()))
+        }
+    }
+
+    #[diplomat::opaque]
+    pub struct Hwnd(pub isize);
+
+    impl Hwnd {
+        #[cfg(target_pointer_width = "32")]
+        pub fn new(hwnd: u32) -> Result<Box<Hwnd>, Box<IronRdpError>> {
+            isize::try_from(hwnd)
+                .map(Hwnd)
+                .map(Box::new)
+                .map_err(|_| "cannot convert from u64 input".into())
+        }
+
+        #[cfg(target_pointer_width = "64")]
+        pub fn new(hwnd: u64) -> Result<Box<Hwnd>, Box<IronRdpError>> {
+            isize::try_from(hwnd)
+                .map(Hwnd)
+                .map(Box::new)
+                .map_err(|_| "cannot convert from u64 input".into())
+        }
+    }
+
+    #[diplomat::opaque]
+    pub struct CliprdrBackendFactory(pub Box<dyn ironrdp::cliprdr::backend::CliprdrBackendFactory>);
+
+    impl CliprdrBackendFactory {
+        pub fn build_cliprdr(&self) -> Box<Cliprdr> {
+            let backend = self.0.build_cliprdr_backend();
+            let cliprdr = ironrdp::cliprdr::Cliprdr::new(backend);
+            Box::new(Cliprdr(Some(cliprdr)))
+        }
+    }
+
+    #[diplomat::opaque]
+    pub struct Cliprdr(pub Option<ironrdp::cliprdr::Cliprdr<Client>>);
+
+    #[diplomat::opaque]
+    pub struct CliprdrReference<'a>(pub &'a ironrdp::cliprdr::Cliprdr<Client>);
```

Comment:
Oof, I don't like the idea of passing the Rust reference to dotnet garbage-collected object and keeping it as a public API -- we may make a shoot in our legs with this someday 😢.

Could we avoid this? maybe by moving `CliprdrReference` methods to `Connection` instead without construction or intermediate "Reference" object?

cc @irvingoujAtDevolution @CBenoit

Reply:
@pacmancoder This should be fine theoretically. It’s currently not sound, but I plan to fix the C# backend for keeping track of such borrows. Diplomat model allows because this can be made safe. They’ve done it for their JavaScript backend for example.

Reply:
Updated, we now hide the reference behind the active stage.